README.md
HistoryPlot
===========

Graph plot script for history data


ChangeLog
/*
 * HistoryPlot
 *
 *
 * Copyright (c) 2014-2015 Advanced Institute for Computational Science, RIKEN. 
 * All rights reserved.
 *
 */


REVISION HISTORY
================

--------------------------
2015-06-05
- enhancement
- add tab and unicode white as a delimiter
- use Konatsu font as default for all platforms


--------------------------
2015-01-16
- move Konatu.tff to src/ [0.7.2]

--------------------------
2014-12-31
- userguide_jp.pdf [0.7.1]
- add range parameter, Konatu.ttf and eps & pdf images [0.7.0]
- initial commit [0.6.0]


src/FileIO.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
# Name:
# Purpose:
# Author:      zrm
# Created:     11/06/2014
# Copyright: AICS, RIKEN
# Licence: New BSD (2-caluse)
#-------------------------------------------------------------------------------

import numpy as np
import sys

#----------------- HardCoded ------------------------------
#以下の定義は、入力ファイルのフォマットによって、HardCodedされます。
the_str_data_begin = 'Detail of Timing Statistics'

g_keys = ['call', 'accm[s]', 'accm[%]', 'waiting[s]',
          'accm/call[s]', 'flop|msg', 'speed']
g_vals = [0, 1, 2, 3, 4, 5, 6]
g_dict = dict(zip(g_keys, g_vals))

g_s = [7,  20, 34, 42, 56, 70, 82]
g_e = [20, 34, 42, 56, 70, 82, 91]

import ParamDef as param
#--------------------------------------------------------------
def utf(any_str):
        if param.G_DECODE_STR == 1 : new_str = any_str.decode('utf-8')
        else: new_str = any_str
        return new_str

#--------------------------------------------------------------
def load_data(filename):

        my_dict = read_file(filename)

        print utf('\n 入力ファイル名： \n'), filename
        print utf('\n Data Retained in Python Dictionary. len = '), len(my_dict), '\n\n'

        for func_name in my_dict:
                #print utf('\nfunction name ='), func_name
                sequence_vec = my_dict[func_name]
                #print utf(' '), len(sequence_vec)
                #for data in sequence_vec:
                #        print data

        #Giving function name and value name, we can get array from my_dict
        #print utf('\ntry to get array by given function name and value name')

        n_funcs = len(sequence_vec)
        return my_dict

#--------------------------------------------------------------
def read_file(filename):
        my_dict = {}
        sequence_vec = []

        f = open(filename)
        if not f: return output

        line = f.readline()
        while not the_str_data_begin in line:
                line = f.readline()
        print utf('\nRead Data: '), the_str_data_begin

        line = f.readline() # get the empty line

        while True:
                func_name  = f.readline()
                if not func_name: break
                func_name = func_name.rstrip('\n')

                #print utf('\nfunction name = '), utf('['), func_name, utf(']\n')
                unicode(func_name).strip()
                func_name.strip()
                #print utf('\nfunction name = '), utf('['), func_name, utf(']\n')

                dummy_line = f.readline() # get the empty line "   call accm[s] accm[%]..."

                line = f.readline()
                if not line: break

                while '#' in line:
                        values = get_data_from_line(line)
                        #print utf('line  :'), line, utf('return:'), values
                        sequence_vec.append(values)

                        line = f.readline()
                        if not line: break

                my_dict[func_name] = np.array(sequence_vec)
                del sequence_vec[:]
        f.close()

        return my_dict


#--------------------------------------------------------------
def get_data_from_line(line):
        array = []

        #values = line.strip().split()
        #for key in g_keys:
        #        array.append( values[ g_dict2[key] ] )

        #　g_□□　はプログラム先頭に定義されているグローバルリストや辞書などです
        ii = 0
        for key in g_keys:
                i = g_s[ii]
                j = g_e[ii]
                str = line[i:j]
                array.append( float(str) )
                ii = ii+1

	return array

###--------------------------------------------------------------
##def arrays_from_file(filename):
##	output = []
##	with open(filename, 'r') as infile:
##		for line in infile:
##			line = np.array(line.strip().split(), dtype=np.float)
##			output.append(line)
##	return output



src/ParamDef.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
# Name:
# Purpose:
# Author:      zrm
# Created:     11/06/2014
# Copyright: AICS, RIKEN
# Licence: New BSD (2-caluse)
#-------------------------------------------------------------------------------

ANY_VAL = 101

#日本語の文字化けを防止するために、これを１にして下さい。
#出力先がファイルで、エンコードエラーを起す時に、これを0にして下さい。
G_DECODE_STR    = 1

#プラットフォームによって、一つを１にして下さい。
G_WINDOWS       = 0
G_MACOSX        = 1
G_LINUX         = 0
G_LINUX_K       = 0
G_LINUX_FOCUS   = 0

#Please use 'Agg' If it issues following error:
#"no display name and no $DISPLAY environment"
#Agg is a non-interactive backend, it won't display
#on the screen, Show() does not work, it saves to files.

if G_LINUX_K == 1 : G_USE_AGG = 1
else:               G_USE_AGG = 0

#global ANY_VAL
#global G_WINDOWS
#global G_MACOSX
#global G_LINUX
#global G_LINUX_K
#global G_USE_AGG


src/Quantile.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
# Name:
# Purpose:
# Author:      Ernesto P. Adorio
# Created:
# Copyright:
# Licence:
#-------------------------------------------------------------------------------

import numpy as np
import sys

import ParamDef as param
#--------------------------------------------------------------
def utf(any_str):
        if param.G_DECODE_STR == 1 : new_str = any_str.decode('utf-8')
        else: new_str = any_str
        return new_str
      
#-----------------------------------------------------------------------------
"""
箱ひげ図（はこひげず、箱髭図、box plot）

                            +-----+-+
  *           o     |-------|   + | |---|
                            +-----+-+

+---+---+---+---+---+---+---+---+---+---+   number line
0   1   2   3   4   5   6   7   8   9  10

このデータセット（値は図から読み取れる概略値とする）から、次のことが分かる:
    最小値（min） = 5。
    第1四分位点（Q1） = 7。
    中央値（第2四分位点、Med） = 8.5。
    第3四分位点（Q3） = 9。
    最大値（max） = 10。
    平均値 = 8。
    IQR(interquartile range) = Q3-Q1 = 2
    3.5という値は"軽度の"外れ値、つまりQ1よりも 1.5×IQR から 3×IQR だけ下にある。
    0.5という値は"極端な"外れ値、つまりQ1よりも 3×IQR 以上下にある。
    外れ値以外の最小値は5。
    データは左に歪んでいる（負の歪度）。
"""
# 箱ひげ図(BoxPlot）をプロットするため、データをソートしています。
# ソートする候補のデータは、この配列に格納しています。
g_Values = [0.0,   0.0,   0.0,   0,0, 0.0,   0.0,   0.0,   0,0]

MED_VAL = 0
MIN_VAL = 1
MAX_VAL = 2
DIF_VAL = 3
Q1_VAL  = 4
Q3_VAL  = 5
AVG_VAL = 6
IQR_VAL = 7

g_ValueName = ['', '', '', '', '', '', '', '']
g_ValueName[MED_VAL] = '中央値(Q2)'         #median (Q2)
g_ValueName[MIN_VAL] = '最小値'             #minimum
g_ValueName[MAX_VAL] = '最大値'             #maximum
g_ValueName[DIF_VAL] = '最大最小差値'       #range
g_ValueName[Q1_VAL]  = '第1四分位数(Q1)'    #first quartile (Q1)
g_ValueName[Q3_VAL]  = '第3四分位数(Q3)'    #third quartile (Q3)
g_ValueName[AVG_VAL] = '平均値'             #average
g_ValueName[IQR_VAL] = '四分位数範囲(IQR)'  #Interquartile range (IQR)

# 配列データの分位数ルゴリズム
#g_qtype = 1 # inverse empirical distrib.function., R type 1
#g_qtype = 2 # similar to type 1, averaged, R type 2
#g_qtype = 3 # nearest order statistic,(SAS) R type 3
#g_qtype = 4 # California linear interpolation, R type 4
#g_qtype = 5 # hydrologists method, R type 5
#g_qtype = 6 # mean-based estimate(Weibull method), (SPSS,Minitab), type 6
g_qtype = 7 # mode-based method,(S, S-Plus), R type 7
#g_qtype = 8 # median-unbiased ,  R type 8
#g_qtype = 9 # normal-unbiased, R type 9.

def calculate_Median_Q1_Q3( array):
  g_Values[MED_VAL]=quantile(array, 0.50, g_qtype)   #median
  g_Values[Q1_VAL ]=quantile(array, 0.25, g_qtype)   #first quartile (Q1)
  g_Values[Q3_VAL ]=quantile(array, 0.75, g_qtype)   #third quartile (Q3)
  g_Values[IQR_VAL]=g_Values[Q3_VAL]-g_Values[Q1_VAL]#Interquartile range (IQR)
  g_Values[MIN_VAL]=np.amin(array)                   #minmum
  g_Values[MAX_VAL]=np.amax(array)                   #maxmum
  g_Values[AVG_VAL]=np.average(array)                #average
  g_Values[DIF_VAL]=g_Values[MAX_VAL]-g_Values[MIN_VAL] #range

#-----------------------------------------------------------------------------

"""
File	quantile.py
Desc	computes sample quantiles
Author  Ernesto P. Adorio, PhD.	 UPDEPP (U.P. at Clarkfield)
Version 0.0.1 August 7. 2009
"""
from math import modf, floor
def quantile(x, q,  qtype = 7, issorted = False):
	"""
	Args:
	   x - input data
	   q - quantile
	   qtype - algorithm
	   issorted- True if x already sorted.
	Compute quantiles from input array x given q.For median,
	specify q=0.5.
	References:
	   http://reference.wolfram.com/mathematica/ref/Quantile.html
	   http://wiki.r-project.org/rwiki/doku.php?id=rdoc:stats:quantile
	Author:
	Ernesto P.Adorio Ph.D.
	UP Extension Program in Pampanga, Clark Field.
	"""
	if not issorted:
		y = sorted(x)
	else:
		y = x
	if not (1 <= qtype <= 9):
	   return None  # error!
	# Parameters for the Hyndman and Fan algorithm
	abcd = [(0,   0, 1, 0), # inverse empirical distrib.function., R type 1
		(0.5, 0, 1, 0), # similar to type 1, averaged, R type 2
		(0.5, 0, 0, 0), # nearest order statistic,(SAS) R type 3
		(0,   0, 0, 1), # California linear interpolation, R type 4
		(0.5, 0, 0, 1), # hydrologists method, R type 5
		(0,   1, 0, 1), # mean-based estimate(Weibull method), (SPSS,Minitab), type 6
		(1,  -1, 0, 1), # mode-based method,(S, S-Plus), R type 7
		(1.0/3, 1.0/3, 0, 1), # median-unbiased ,  R type 8
		(3/8.0, 0.25, 0, 1)   # normal-unbiased, R type 9.
	       ]
	a, b, c, d = abcd[qtype-1]
	n = len(x)
	g, j = modf( a + (n+b) * q -1)
	if j < 0:
		return y[0]
	elif j >= n:
		return y[n-1]   # oct. 8, 2010 y[n]???!! uncaught  off by 1 error!!!
	j = int(floor(j))
	if g ==  0:
	   return y[j]
	else:
	   return y[j] + (y[j+1]- y[j])* (c + d * g)
def Test():
	x = [11.4, 17.3, 21.3, 25.9, 40.1, 50.5, 60.0, 70.0, 75]
	for qtype in range(1,10):
		print qtype, quantile(x, 0.35, qtype)
if __name__ == "__main__":
	Test()
"""
When the test code runs, it outputs
    1 25.9
    2 25.9
    3 21.3
    4 21.99
    5 24.29
    6 23.6
    7 24.98
    8 24.06
    9 24.1175
This matches the output of the following R code:
x <- c(11.4, 17.3, 21.3, 25.9, 40.1, 50.5, 60.0, 70.0, 75)
for (i in seq(1,9)) { print(c( i, quantile(x, 0.35, type=i))) }
"""



src/PlotOption.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
# Name:
# Purpose:
# Author:      zrm
# Created:     11/06/2014
# Copyright: AICS, RIKEN
# Licence: New BSD (2-caluse)
#-------------------------------------------------------------------------------

class PlotOption(object):
    def __init__(self):
        self.figwidth = 10
        self.figheight= 10
        self.figdpi   = 80
        self.facecolor = 'w'
        self.edgecolor = 'k'
        self.drawcolor = 'k'
        self.drawwidth = 0.6
        self.figalpha = 1.0
        self.figcolor = 'w'

        self.title = ''
        self.titlecolor = 24
        self.width = 20
        self.titlefont  = 18

        self.labelcolor = 'blue'
        self.labelfont = 18
        self.xlabel= 'X'
        self.ylabel= 'Y'

        self.tickcolor = 'black'
        self.xtitlerot  = 0
        self.ytitlerot  = 0
        self.tickfont   = 16

        self.textflag = 0
        self.textposx = 0.5 # pct
        self.textposy = 0.5 # pct
        self.text = 'text'
        self.textfont = 16
        self.textcolor= 'red'
        self.textalpha= 0.5

        self.xrange = 0
        self.xmin = 0.0;
        self.xmax = 0.0;
        self.yrange = 0
        self.ymin = 0.0;
        self.ymax = 0.0;

        self.pngname = ''
        self.logscale = 0
    
    def set_logscale(self, zero_or_not):
        self.logscale = zero_or_not

    def set_xrange(self, amin, amax):
        self.xrange = 1;
        self.xmin = amin
        self.xmax = amax

    def set_yrange(self, amin, amax):
        self.yrange = 1;
        self.ymin = amin
        self.ymax = amax

    def set_pngname(self, pngname):
        self.pngname = pngname

    def set_tick(self, color, font, rotx, roty):
        self.tickcolor = color
        self.xtitlerot = rotx
        self.ytitlerot = roty
        self.tickfont  = font

    def set_text(self, text, posx=0.5, posy=0.5, font=16, color='k', color_bg='green', alpha=0.5):
        self.textflag = 1
        self.textposx = posx
        self.textposy = posy
        self.text = text
        self.textfont = font
        self.textcolor= color
        self.textcolor_gb = color_bg
        self.textalpha= alpha

    def set_draw(self, drawcolor, edgecolor, drawwidth):
        self.drawcolor = drawcolor
        self.edgecolor = edgecolor
        self.drawwidth = drawwidth

    def set_figcolor(self, color, alpha):
        self.figalpha = alpha
        self.figcolor = color

    def set_figsize(self, width, height):
        self.figwidth = width
        self.figheight= height

    def set_title(self, title, font, color):
        self.title = title
        self.titlefont = font
        self.titlecolor = color

    def set_label(self, xlabel, ylabel, font, color):
        self.xlabel= xlabel
        self.ylabel= ylabel
        self.labelfont = font
        self.labelcolor = color




src/FileIOColumnData.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
# Name:
# Purpose:
# Author:      zrm
# Created:     11/06/2014
# Copyright: AICS, RIKEN
# Licence: New BSD (2-caluse)
#-------------------------------------------------------------------------------

import numpy as np
import sys

import ParamDef as param
#--------------------------------------------------------------
def utf(any_str):
        if param.G_DECODE_STR == 1 : new_str = any_str.decode('utf-8')
        else: new_str = any_str
        return new_str

#--------------------------------------------------------------
def replace_many_spaces_by_one(my_string):
        my_string = my_string.strip()
        while '  ' in my_string:
            my_string = my_string.replace('  ', ' ')
        return my_string

#--------------------------------------------------------------
def read_file(filename, data_block_name, column_names):

        data_dict = {}
        sequence_vec = []
        
        f = open(filename)
        if not f: return output

        # read lines until the one we want.
        line = f.readline()
        while not data_block_name in line:
                line = f.readline()
        print utf('Found the Named Data Block: '), data_block_name

        #Note:  It is no allowed to have blank line between
        #       data_block_name and column name line
        
        #get column name line
        column_name_line = f.readline()
        if not column_name_line: return output
        
        #print utf('before: column_name_line  :'), column_name_line
        column_name_line = replace_many_spaces_by_one(column_name_line)
        #print utf('after : column_name_line  :'), column_name_line

        #put column names into a string array and a corresponding index array
        #then, we construct a dictionary for finding index by column name.
        column_names = column_name_line.strip().split()
        #print utf('column_names  :'), column_names
        
        step=1
        while True:
                line = f.readline()
                if not line: break
                
                line = replace_many_spaces_by_one(line)

                values = line.strip().split()
                #print utf('values:'), values

                sequence_vec.append(values)
                #print utf('sequence_vec:'), sequence_vec
                        
                data_dict[step] = np.array(sequence_vec)
                
                del sequence_vec[:]
                
                step = step+1
                
        f.close()

        return data_dict, column_names


#--------------------------------------------------------------
def load_data(filename, data_block_name, column_names):

        print utf('入力ファイル名：'), filename
        
        data_dict, column_names = read_file(filename, data_block_name, column_names)
        
        print utf('Data Retained in Data Array. len = '), len(data_dict)

        '''
        for step in data_dict:
                sequence_vec = data_dict[step]

                print utf('sequence_vec= '), sequence_vec

                #print utf(' '), len(sequence_vec)
                #for data in sequence_vec:
                #        print data
        '''

        return data_dict, column_names




src/PlotMatplotlib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
# Name:
# Purpose:
# Author:      zrm
# Created:     11/06/2014
# Copyright: AICS, RIKEN
# Licence: New BSD (2-caluse)
#-------------------------------------------------------------------------------

import ParamDef as param
#--------------------------------------------------------------
def utf(any_str):
        if param.G_DECODE_STR == 1 : new_str = any_str.decode('utf-8')
        else: new_str = any_str
        return new_str
#--------------------------------------------------------------

import matplotlib

#Please use 'Agg' If it issues following error
#"no display name and no $DISPLAY environment"
#
#Agg is a non-interactive backend, meaning it won't display
#on the screen, Show() does not work, it only save to files.

if param.G_USE_AGG == 1: matplotlib.use('Agg')

import matplotlib.font_manager as font_manager
import matplotlib.patches as patches
import matplotlib.cbook as cbook
import matplotlib.pyplot as plt

import numpy as np
import sys

from matplotlib.patches import Polygon

from Quantile import *

# 日本語fontの設定
g_fontsize = 18

import os
bdir = os.path.dirname(sys.argv[0])
font_path = os.path.join(bdir, 'Konatu.ttf')
"""
if   param.G_WINDOWS == 1:
        # for Windows
        #ttfフォントでない場合、pdf と eps 作成できません、'konatu.ttf'を使用します。
        #font_path = 'C:/WINDOWS/Fonts/meiryo.ttc' 
        font_path = 'konatu.ttf'
        #  -----------------------------------------------------------------
        # 『小夏』("Konatu") Copyright (C) 2002〜 桝田道也 All rights reserved.
        #  http://avoidnotes.org/urlmemo/cache/20050722191344.html#about
        # 『小夏』のライセンスはver.26（2010-01）までは CC-BY-SA 3.0 でしたが、
        #  ver.20121218 から The MIT License に変更しました。 
        #  -----------------------------------------------------------------
elif param.G_MACOSX == 1:
        # for Mac
        font_path = '/Library/Fonts/Osaka.ttf'
elif param.G_LINUX_K == 1 :
        # for Linux of K-Computer
        font_path = '/Please/Specify/ProperFont.ttf'
elif param.G_LINUX_FOCUS == 1 :
        # for Linux of FOCUS
        font_path = '/Please/Specify/ProperFont.ttf'
else:
        # for CentOS
        font_path = '/usr/share/fonts/ja/TrueType/kochi-gothic-subst.ttf'
"""

g_font_prop = font_manager.FontProperties(fname=font_path)
g_font_prop.set_style('normal')
g_font_prop.set_weight('light')
g_font_prop.set_size(g_fontsize)

#--------------------------------------------------------------
def set_rect_color(rect, color, alpha):
        rect.set_facecolor(color)
        rect.set_alpha(alpha)

def set_tick_xlabel(ax, color, font, rot):
        for label in ax.get_xticklabels():
            label.set_color(color)
            label.set_rotation(rot)
            label.set_fontsize(font)

def set_tick_ylabel(ax, color, font, rot):
        for label in ax.get_yticklabels():
            label.set_color(color)
            label.set_rotation(rot)
            label.set_fontsize(font)
	
#--------------------------------------------------------------
def draw_barplot(x_data, y_data, opt, graph_type, finalize, mute):

        #　図の縦、横を定義する
        fig = plt.figure( num=0,                                \
                          figsize=(opt.figwidth, opt.figheight),\
                          dpi=opt.figdpi,                       \
                          facecolor=opt.facecolor,              \
                          edgecolor=opt.edgecolor,              \
                         )

        #　図を作成する
        handles = 0

        if graph_type == 'bar' :
                handles, = plt.bar(x_data, y_data,             \
                        facecolor=opt.drawcolor,    \
                        edgecolor=opt.edgecolor,    \
                        width=opt.drawwidth,        \
                        label='----',               \
                        align='center'              \
                        )
        else:
                handles, = plt.plot(x_data, y_data,
                         color=opt.drawcolor,
                         linewidth=opt.drawwidth,
                         linestyle='-'
                         )

        #　図の枠範囲の色と透明度を指定する
        set_rect_color( fig.patch, opt.figcolor, opt.figalpha )

        #　図のタイトルを設定する
        plt.title(  opt.title,
                    fontsize=opt.titlefont,
                    color=opt.titlecolor,
                    fontproperties=g_font_prop
                 )

        #　縦、横軸の範囲、ラベル、色、フォントサイズを設定する
        plt.xlabel(opt.xlabel, fontsize=opt.labelfont,
                   color=opt.labelcolor, fontproperties=g_font_prop)

        plt.ylabel(opt.ylabel, fontsize=opt.labelfont,
                   color=opt.labelcolor, fontproperties=g_font_prop)

        x_min = np.amin(x_data)
        x_max = np.amax(x_data)

        y_min = np.amin(y_data)
        y_max = np.amax(y_data) * 1.1 # make margin for showing text at top

        if opt.xrange > 0 :
            plt.xlim(xmin=opt.xmin)
            plt.xlim(xmax=opt.xmax)
            x_min = opt.xmin
            x_max = opt.xmax

        if opt.yrange > 0 :
            plt.ylim(ymin=opt.ymin)
            plt.ylim(ymax=opt.ymax)
            y_min = opt.ymin
            y_max = opt.ymax

        #　図の一部をハイライトする方法
        opacity = 0.001
        #縦ハイライト範囲
        plt.axvspan(x_max*0.25, x_max*0.75, color='red',  alpha=opacity)
        #横ハイライト範囲
        plt.axhspan(y_max*0.25, y_max*0.75, color='blue', alpha=opacity)

        #　図の範囲の色、軸の刻みのラベルの色、フォント、回転を指定する
        #ax = fig.add_subplot(111, axisbg='darkslategray')
        ax = fig.add_subplot(111)

        # 画像をグラフに入れる　テスト！
        #image_file = cbook.get_sample_data('C:\\_ZrmData\\ffv_ono_graph\\logo.png')
        #image = plt.imread(image_file)
        #im = ax.imshow(image)
        #patch = patches.Circle((260, 200), radius=200, transform=ax.transData)
        #im.set_clip_path(patch)

        set_rect_color( ax.patch, 'white', 1.0 )

        set_tick_xlabel(ax, opt.tickcolor, opt.tickfont, opt.xtitlerot)
        set_tick_ylabel(ax, opt.tickcolor, opt.tickfont, opt.ytitlerot)

        #handles, labels = ax.get_legend_handles_labels()
        #ax.legend(handles, labels)

        #ax.fill(x_data,y_data, 'b' )

        if opt.logscale > 0 :
                ax.set_yscale('log')

        #　文字箱を作成する
        if opt.textflag > 0 :
                g_font_prop.set_size(opt.textfont)
                ax.text( x_max*opt.textposx,                    \
                         y_max*opt.textposy,                    \
                         opt.text,                              \
                         style='italic',                        \
                         fontproperties=g_font_prop,            \
                         fontsize=opt.textfont,                 \
                         bbox={'facecolor':opt.textcolor,       \
                               'alpha':opt.textalpha,           \
                               'pad':opt.textfont}              \
                         )
                g_font_prop.set_size(g_fontsize)

        # savefig() must be called before show(), otherwise, blank image.
        if len(opt.pngname) >=8 : # The shortest pngname C:\a.png
                plt.savefig(opt.pngname, bbox_inches="tight", pad_inches=0.15)

        #plt.legend(loc='upper right')

        if finalize > 0 :
                if mute <= 0 :  plt.show()
                else :          plt.clf()

        return handles

#--------------------------------------------------------------
def draw_boxplot_a(list_of_data_array, xticks, xticklabels, rot, opt, mute):

        fig, ax1 = plt.subplots(figsize=(opt.figwidth, opt.figheight))

        fig.canvas.set_window_title(opt.title)

        plt.subplots_adjust(left=0.075, right=0.95, top=0.9, bottom=0.25)

        bp = plt.boxplot(list_of_data_array, notch=0, sym='+', vert=1, whis=1.5)

        plt.setp(bp['boxes'],    color='black') # 箱の輪郭
        plt.setp(bp['whiskers'], color='blue')  # 中央値の線
        plt.setp(bp['fliers'],   color='lightgrey', marker='+')# 異常値の点

        ax1.yaxis.grid(True, linestyle='-', which='major',
                       color='lightgrey', alpha=0.5)

        ax1.set_axisbelow(True) # グリッド線を図の下に置く

        ax1.set_title(opt.title, fontsize=opt.titlefont,
                      color=opt.titlecolor, fontproperties=g_font_prop)

        ax1.set_xlabel(opt.xlabel, fontsize=opt.labelfont,
                       color=opt.labelcolor, fontproperties=g_font_prop)

        ax1.set_ylabel(opt.ylabel, fontsize=opt.labelfont,
                       color=opt.labelcolor, fontproperties=g_font_prop)

        boxColors = ['darkkhaki','royalblue']
        numBoxes = len(list_of_data_array)
        medians = range(numBoxes)
        for i in range(numBoxes):
            box = bp['boxes'][i]

            boxX = []
            boxY = []
            for j in range(5):
                    boxX.append(box.get_xdata()[j])
                    boxY.append(box.get_ydata()[j])
                    boxCoords = zip(boxX,boxY)

            k = i % 2 # 二つの色を切り替える
            boxPolygon = Polygon(boxCoords, facecolor=boxColors[k], alpha=0.5)

            ax1.add_patch(boxPolygon)

        x_min = 0
        x_max = len(xticks)

        y_min = np.amin(list_of_data_array)
        y_max = np.amax(list_of_data_array)

        if opt.xrange > 0 :
            x_min = opt.xmin
            x_max = opt.xmax
            plt.xlim(xmin=opt.xmin)
            plt.xlim(xmax=opt.xmax)

        if opt.yrange > 0 :
            y_min = opt.ymin
            y_max = opt.ymax
            plt.ylim(ymin=opt.ymin)
            plt.ylim(ymax=opt.ymax)

        plt.xticks(xticks, xticklabels, rotation=75,
                   fontsize=opt.tickfont, color=opt.tickcolor)

        plt.yticks(fontsize=opt.tickfont, color=opt.tickcolor)

        # 中央値を計算し、図の上に中央値を表示する
        b_show_median = 1
        if b_show_median > 0 :
                ii=0
                for array in list_of_data_array:
                    medians[ii] = quantile(array, 0.5, g_qtype)
                    ii=ii+1

                top = y_max
                pos = np.arange(numBoxes)+1
                upperLabels = [str(np.round(s, 2)) for s in medians]
                weights = ['bold', 'semibold']
                for tick,label in zip(range(numBoxes),ax1.get_xticklabels()):
                   k = tick % 2
                   ax1.text(    pos[tick], top-(top*0.05),
                                upperLabels[tick],
                                horizontalalignment='center',
                                size='x-small',
                                weight=weights[k],
                                color=boxColors[k])

        if opt.textflag > 0 :
                g_font_prop.set_size(opt.textfont)
                plt.figtext(opt.textposx,
                            opt.textposy,
                            opt.text ,
                            fontproperties=g_font_prop,
                            backgroundcolor='lightgrey',
                            color=opt.textcolor,
                            weight='roman',
                            fontsize=opt.textfont,
                            bbox={'facecolor':opt.textcolor_gb,
                                  'alpha':opt.textalpha,
                                  'pad':opt.textfont}
                            )
                g_font_prop.set_size(g_fontsize)

        # savefig() must be called before show(), otherwise, blank image.
        if len(opt.pngname) >=8 : # The shortest pngname C:\a.png
                plt.savefig(opt.pngname, bbox_inches="tight", pad_inches=0.15)

        if mute <= 0 : plt.show()
        else :         plt.clf()

        return

def draw_boxplot(list_of_data_array, xticks, xticklabels, rot, opt, mute):

        draw_boxplot_a(list_of_data_array, xticks, xticklabels, rot, opt, mute)

        return

        #　図の縦、横を定義する
        fig = plt.figure( num=0,                                \
                          figsize=(opt.figwidth, opt.figheight),\
                          dpi=opt.figdpi,                       \
                          facecolor=opt.facecolor,              \
                          edgecolor=opt.edgecolor,              \
                         )

        plt.boxplot( list_of_data_array, 0, '' )

        #そうしないと、横軸のTickLabelがははみ出していまう。
        plt.tight_layout(pad=15)

        #　図の枠範囲の色と透明度を指定する
        set_rect_color( fig.patch, opt.figcolor, opt.figalpha )

        #　図のタイトルを設定する
        plt.title(opt.title, fontsize=opt.titlefont,
                  color=opt.titlecolor, fontproperties=g_font_prop)

        #　縦、横軸の範囲、ラベル、色、フォントサイズを設定する
        plt.xlabel(opt.xlabel, fontsize=opt.labelfont,
                   color=opt.labelcolor, fontproperties=g_font_prop)

        plt.ylabel(opt.ylabel, fontsize=opt.labelfont,
                   color=opt.labelcolor, fontproperties=g_font_prop)

        x_min = 0
        x_max = len(xticks)

        y_min = np.amin(list_of_data_array)
        y_max = np.amax(list_of_data_array)

        plt.xticks(xticks, xticklabels, rotation=75,
                   fontsize=opt.tickfont, color=opt.tickcolor)

        plt.yticks(fontsize=opt.tickfont, color=opt.tickcolor)

        if opt.xrange > 0 :
            x_min = opt.xmin
            x_max = opt.xmax
            plt.xlim(xmin=opt.xmin)
            plt.xlim(xmax=opt.xmax)

        if opt.yrange > 0 :
            y_min = opt.ymin
            y_max = opt.ymax
            plt.ylim(ymin=opt.ymin)
            plt.ylim(ymax=opt.ymax)

        #　図の範囲の色、軸の刻みのラベルの色、フォント、回転を指定する
        ax = fig.add_subplot(111)
        set_rect_color( ax.patch, 'white', 1.0 )

        #　文字箱を作成する
        if opt.textflag > 0 :
                g_font_prop.set_size(opt.textfont)
                ax.text( x_max*opt.textposx,                    \
                         y_max*opt.textposy,                    \
                         opt.text,                              \
                         style='italic',                        \
                         fontproperties=g_font_prop,            \
                         fontsize=opt.textfont,                 \
                         bbox={'facecolor':opt.textcolor,       \
                               'alpha':opt.textalpha,           \
                               'pad':opt.textfont}              \
                         )
                g_font_prop.set_size(g_fontsize)

        # マージンをセットする
        b_set_margin = 0
        if b_set_margin > 0 :
                plot_margin_x = x_max * 0.02
                plot_margin_y = y_max * 0.02
                x0, x1, y0, y1 = plt.axis()
                plt.axis((x0 - plot_margin_x,
                          x1 + plot_margin_x,
                          y0 - plot_margin_y,
                          y1 + plot_margin_y))

        # savefig() must be called before show(), otherwise, blank image.
        if len(opt.pngname) >=8 : # The shortest pngname C:\a.png
                plt.savefig(opt.pngname, bbox_inches="tight", pad_inches=0.15)

        if mute <= 0 : plt.show()
        else :          plt.clf()
